src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -397,27 +370,426 @@ CollectionExpressionTypeKind.Span or
+                throw ExceptionUtilities.UnexpectedValue(node);
+            }
 
-            AddPlaceholderReplacement(addElementPlaceholder, _factory.Local(iterationVariable));
+            if (numberIncludingLastSpread == 0)
+            {
+                int knownLength = elements.Length;
+                if (knownLength == 0)
+                {
+                    return CreateEmptyArray(syntax, arrayType);
+                }
 
-            var rewrittenBody = VisitStatement(initializer.AddMethodInvocation);
-            Debug.Assert(rewrittenBody is { });
+                var initialization = new BoundArrayInitialization(
+                    syntax,
+                    isInferred: false,
+                    elements.SelectAsArray(static (element, rewriter) => rewriter.VisitExpression(element), this));
+                return new BoundArrayCreation(
+                    syntax,
+                    ImmutableArray.Create<BoundExpression>(
+                        new BoundLiteral(
+                            syntax,
+                            ConstantValue.Create(knownLength),
+                            _compilation.GetSpecialType(SpecialType.System_Int32))),
+                    initialization,
+                    arrayType)
+                { WasCompilerGenerated = true };
+            }
 
-            RemovePlaceholderReplacement(addElementPlaceholder);
+            BoundAssignmentOperator assignmentToTemp;
+            var localsBuilder = ArrayBuilder<BoundLocal>.GetInstance();
+            var sideEffects = ArrayBuilder<BoundExpression>.GetInstance();
+
+            RewriteCollectionExpressionElementsIntoTemporaries(elements, numberIncludingLastSpread, localsBuilder, sideEffects);
+
+            // int index = 0;
+            BoundLocal indexTemp = _factory.StoreToTemp(
+                _factory.Literal(0),
+                out assignmentToTemp,
+                isKnownToReferToTempIfReferenceType: true);
+            localsBuilder.Add(indexTemp);
+            sideEffects.Add(assignmentToTemp);
+
+            // ElementType[] array = new ElementType[N + s1.Length + ...];
+            BoundLocal arrayTemp = _factory.StoreToTemp(
+                new BoundArrayCreation(syntax,
+                    ImmutableArray.Create(GetKnownLengthExpression(elements, numberIncludingLastSpread, localsBuilder)),
+                    initializerOpt: null,
+                    arrayType),
+                out assignmentToTemp,
+                isKnownToReferToTempIfReferenceType: true);
+            localsBuilder.Add(arrayTemp);
+            sideEffects.Add(assignmentToTemp);
+
+            AddCollectionExpressionElements(
+                elements,
+                arrayTemp,
+                localsBuilder,
+                numberIncludingLastSpread,
+                sideEffects,
+                (ArrayBuilder<BoundExpression> expressions, BoundExpression arrayTemp, BoundExpression rewrittenValue) =>
```

Comment:
On the other hand, the refactoring here (inlining of the local function) seems negative to me as it makes the main body of the method harder to follow (much logic inside the lambda now). Consider keeping original structure.

Reply:
Will keep as is, since there are three separate callers of `AddCollectionExpressionElements()` that each pass a lambda expression with inline method.